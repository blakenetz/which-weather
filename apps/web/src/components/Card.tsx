import {
  Button,
  Card as MuiCard,
  CardActions,
  CardContent,
  Typography,
  Stack,
  Box,
} from "@mui/material";
import { Forecast } from "@server/types";
import { format, isToday, isTomorrow } from "date-fns";

interface CardProps {
  data: Forecast;
}

function formatDate(date: string) {
  let formatString = "ccc";

  if (isToday(date)) {
    formatString = "'Today'";
  } else if (isTomorrow(date)) {
    formatString = "'Tomorrow'";
  }

  return format(date, `${formatString} 'at' h:mm aaaa`);
}

export default function Card({ data }: CardProps) {
  const date = formatDate(data.time);

  return (
    <MuiCard sx={{ minWidth: 300 }}>
      <CardContent sx={{ display: "flex", flexDirection: "column", gap: 1 }}>
        <Stack direction="row">
          {data.icon && (
            <Box
              component="img"
              src={data.icon}
              sx={({ typography }) => ({
                height: `calc(${typography.subtitle1.fontSize} * 2)`,
              })}
            />
          )}
          <Typography variant="subtitle2">{date}</Typography>
        </Stack>
        <div>
          {Array.isArray(data.temperature) ? (
            <div>
              <Typography>High: {data.temperature[0]}째F</Typography>
              <Typography>Low: {data.temperature[1]}째F</Typography>
            </div>
          ) : (
            <Typography>Temp: {data.temperature}째F</Typography>
          )}
          {data.feelsLike && (
            <Typography>Feels like: {data.feelsLike}째F</Typography>
          )}
        </div>
        {data.description && <Typography>{data.description}</Typography>}
        {data.wind && <Typography>{data.wind}</Typography>}
      </CardContent>
      {data.link && (
        <CardActions sx={{ justifyContent: "flex-end" }}>
          <Button size="small" component="a" href={data.link} target="_blank">
            Source
          </Button>
        </CardActions>
      )}
    </MuiCard>
  );
}
