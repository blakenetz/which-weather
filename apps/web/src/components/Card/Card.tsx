import { Air, Thermostat } from "@mui/icons-material";
import {
  Button,
  Card as MuiCard,
  CardActions,
  CardContent,
  Typography,
  CardHeader,
  Avatar,
  Stack,
  styled,
  Box,
  svgIconClasses,
} from "@mui/material";
import { Forecast } from "@server/types";
import { format, isToday, isTomorrow } from "date-fns";
import { startCase } from "lodash";

interface CardProps {
  data: Forecast;
}

function formatDate(date: string) {
  let formatString = "ccc";

  if (isToday(date)) {
    formatString = "'Today'";
  } else if (isTomorrow(date)) {
    formatString = "'Tomorrow'";
  }

  return format(date, `${formatString} 'at' h:mm aaa`);
}

function isAfternoon(date: string) {
  return format(date, "a") === "PM";
}

function Temperature({
  type,
  temp,
}: {
  type?: "high" | "low" | "feelsLike";
  temp: number;
}) {
  return (
    <Typography variant="body2">
      {`${startCase(type ?? "temp")}: `}
      <Typography
        variant="body2"
        sx={{ color: "text.secondary" }}
        component="span"
      >
        {temp}Â°F
      </Typography>
    </Typography>
  );
}

const Section = styled(Box)(({ theme }) => ({
  display: "flex",
  flexDirection: "row",
  alignItems: "center",
  gap: theme.spacing(1),
  [`.${svgIconClasses.root}`]: {
    height: theme.typography.body2.fontSize,
    width: "auto",
  },
}));

export default function Card({ data }: CardProps) {
  const date = formatDate(data.time);
  const isDark = isAfternoon(data.time);

  return (
    <MuiCard
      raised
      sx={{
        minWidth: 300,
        display: "flex",
        flexDirection: "column",
        justifyContent: "space-between",
      }}
    >
      <div>
        <CardHeader
          title={date}
          avatar={data.icon && <Avatar src={data.icon} />}
          sx={{
            alignItems: "flex-start",
            bgcolor: isDark ? "grey.900" : "primary.light",
            color: isDark ? "primary.contrastText" : "text.primary",
          }}
        />

        <CardContent sx={{ display: "flex", flexDirection: "column", gap: 1 }}>
          <Typography variant="body2" sx={{ color: "text.secondary" }}>
            {data.description}
          </Typography>
          <Section>
            <Thermostat />
            <Stack>
              {Array.isArray(data.temperature) ? (
                <div>
                  <Temperature type="high" temp={data.temperature[1]} />
                  <Temperature type="low" temp={data.temperature[0]} />
                </div>
              ) : (
                <Temperature temp={data.temperature} />
              )}
              {data.feelsLike && (
                <Temperature type="feelsLike" temp={data.feelsLike} />
              )}
            </Stack>
          </Section>

          {data.wind && (
            <Section>
              <Air />
              <Typography>{data.wind}</Typography>
            </Section>
          )}
        </CardContent>
      </div>

      {data.link && (
        <CardActions sx={{ justifyContent: "flex-end" }}>
          <Button size="small" component="a" href={data.link} target="_blank">
            Source
          </Button>
        </CardActions>
      )}
    </MuiCard>
  );
}
