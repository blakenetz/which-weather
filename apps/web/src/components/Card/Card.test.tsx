import { expect, test } from "vitest";
import { render } from "@testing-library/react";
import Card from "./Card";

test("Form", async () => {
  const { baseElement } = render(
    <Card
      data={{
        time: new Date(2000, 1, 1).toString(),
        temperature: 100,
        icon: "",
        description: "description",
        wind: "wind",
      }}
    />
  );

  expect(baseElement).toMatchSnapshot();
});
