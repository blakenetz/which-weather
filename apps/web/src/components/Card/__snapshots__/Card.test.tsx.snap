// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Form 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiCard-root css-1ret5sh-MuiPaper-root-MuiCard-root"
      style="--Paper-shadow: 0px 5px 5px -3px rgba(0,0,0,0.2),0px 8px 10px 1px rgba(0,0,0,0.14),0px 3px 14px 2px rgba(0,0,0,0.12);"
    >
      <div
        class="MuiCardHeader-root css-1v278ml-MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-content css-1jkp9bu-MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-16xl4zq-MuiTypography-root"
          >
            Tue at 12:00 am
          </span>
        </div>
      </div>
      <div
        class="MuiCardContent-root css-fmh09u-MuiCardContent-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 css-1wle3ir-MuiTypography-root"
        >
          description
        </p>
        <div
          class="MuiBox-root css-1x7a4h"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="ThermostatIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15 13V5c0-1.66-1.34-3-3-3S9 3.34 9 5v8c-1.21.91-2 2.37-2 4 0 2.76 2.24 5 5 5s5-2.24 5-5c0-1.63-.79-3.09-2-4m-4-2V5c0-.55.45-1 1-1s1 .45 1 1v1h-1v1h1v2h-1v1h1v1z"
            />
          </svg>
          <div
            class="MuiStack-root css-nen11g-MuiStack-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
            >
              Temp: 
              <span
                class="MuiTypography-root MuiTypography-body2 css-1wle3ir-MuiTypography-root"
              >
                100
                Â°F
              </span>
            </p>
          </div>
        </div>
        <div
          class="MuiBox-root css-1x7a4h"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="AirIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M14.5 17c0 1.65-1.35 3-3 3s-3-1.35-3-3h2c0 .55.45 1 1 1s1-.45 1-1-.45-1-1-1H2v-2h9.5c1.65 0 3 1.35 3 3M19 6.5C19 4.57 17.43 3 15.5 3S12 4.57 12 6.5h2c0-.83.67-1.5 1.5-1.5s1.5.67 1.5 1.5S16.33 8 15.5 8H2v2h13.5c1.93 0 3.5-1.57 3.5-3.5m-.5 4.5H2v2h16.5c.83 0 1.5.67 1.5 1.5s-.67 1.5-1.5 1.5v2c1.93 0 3.5-1.57 3.5-3.5S20.43 11 18.5 11"
            />
          </svg>
          <p
            class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
          >
            wind
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;
